packages/table-schema-provider/src/Features.ts

```diff
@@ -0,0 +1,159 @@
+import { propOr, pathOr } from 'ramda';
+import { TableSchema, tablesListSelectors, tableSelectors, tableFieldSelectors, FIELD_TYPE } from '@8base/utils';
+
+type IsTableFeaturesAllowedCommonArgs = {
+  tableName?: string;
+  appName?: string;
+  tableId?: string;
+};
+
+type IsFieldFeaturesAllowedCommonArgs = {
+  tableName?: string;
+  appName?: string;
+  fieldName: string;
+  tableId?: string;
+  fieldId?: string;
+};
+
+const getFieldSchemaFeatures = (
+  { tableName, appName, fieldName, tableId, fieldId }: IsFieldFeaturesAllowedCommonArgs,
+  tablesList: TableSchema[],
+) => {
+  if (!!tableId && !!fieldId) {
+    const table = tablesListSelectors.getTableById(tablesList, tableId);
+    const field = tableSelectors.getFieldById(table, fieldId);
+    return tableFieldSelectors.getSchemaFeatures(field);
+  } else if (!!tableName && !!fieldName) {
+    const table = tablesListSelectors.getTableByName(tablesList, tableName, appName);
+    const field = tableSelectors.getFieldByName(table, fieldName);
+    return tableFieldSelectors.getSchemaFeatures(field);
+  }
+
+  return {};
+};
+
+const getFieldDataFeatures = (
+  { tableName, appName, fieldName, tableId, fieldId }: IsFieldFeaturesAllowedCommonArgs,
+  tablesList: TableSchema[],
+) => {
+  if (!!tableId && !!fieldId) {
+    const table = tablesListSelectors.getTableById(tablesList, tableId);
+    const field = tableSelectors.getFieldById(table, fieldId);
+    return tableFieldSelectors.getDataFeatures(field);
+  } else if (!!tableName && !!fieldName) {
+    const table = tablesListSelectors.getTableByName(tablesList, tableName, appName);
+    const field = tableSelectors.getFieldByName(table, fieldName);
+    return tableFieldSelectors.getDataFeatures(field);
+  }
+
+  return {};
+};
+
+const getTableSchemaFeatures = (
+  { tableName, appName, tableId }: IsTableFeaturesAllowedCommonArgs,
+  tablesList: TableSchema[],
+) => {
+  if (!!tableId) {
+    const table = tablesListSelectors.getTableById(tablesList, tableId);
+    return tableSelectors.getSchemaFeatures(table);
+  } else if (!!tableName) {
+    const table = tablesListSelectors.getTableByName(tablesList, tableName, appName);
+    return tableSelectors.getSchemaFeatures(table);
+  }
+
+  return {};
+};
+
+const getTableDataFeatures = (
+  { tableName, appName, tableId }: IsTableFeaturesAllowedCommonArgs,
+  tablesList: TableSchema[],
+) => {
+  if (!!tableId) {
+    const table = tablesListSelectors.getTableById(tablesList, tableId);
+    return tableSelectors.getDataFeatures(table);
+  } else if (!!tableName) {
+    const table = tablesListSelectors.getTableByName(tablesList, tableName, appName);
+    return tableSelectors.getDataFeatures(table);
+  }
+
+  return {};
+};
+
+const Features = {
```

Comment:
Why is this variable uppercased?

Reply:
Because it used like `Features.data.isUpdateFieldAllowed(..)`. But I don't know. Maybe better to write `features.data.isUpdateFieldAllowed(..)`.

Reply:
It's strange for me, because `Features` isn't a class. But maybe we can consider them as static class.